rules/fetch-data-nextjs/rule.md

```diff
@@ -0,0 +1,165 @@
+  const res = await fetch(`https://jsonplaceholder.typicode.com/posts/${id}`);
+  const post = await res.json();
+
+  return {
+    props: { post }
+  };
+}
+
+export default function Page(props) {
+  return (
+    <div>
+      <h2>{props.post.title}</h2>
+      <p>{props.post.body}</p> 
+    </div>
+  )
+}
+
+```
+
+### Incremental Static Regeneration (Hybrid)
+
+Server-side generation is great for SEO, however if we have a data source that may change between builds, we may need to regenerate the static data generated at build time. With incremental static regeneration (ISR), we can add an expiry time for static data, and Next.js will automatically refetch the data if the expiry time has been reached. However, this will not block the current request where it has noticed that the data has expired with a long loading time - it will fetch the data only for the next request.
+
+```tsx
+
+export const getStaticProps = () => {
+  const res = await fetch(`https://jsonplaceholder.typicode.com/comments`);
+  const comments = await res.json();
+
+  return {
+    props: { comments },
+    revalidate: 60
+  };
+}
+
+```
+
+This means that if 60 seconds or more has passed after the last time `getStaticProps` was run and the user makes a request to the page, it will rerun the code inside getStaticProps, and render the newly fetched data for the next page visitor.  
+
+### getServerSideProps
+
+`getServerSideProps` allows for server-side fetching of data on each request from the client, which makes it great for fetching of dynamic data. It can also be used for secured data, as the code within the function only runs on the server. 
+
+The below example shows an example of how we can use `getServerSideProps` to fetch data. Upon each user's request, the server will fetch the list of posts and pass it as props to the page.  
+
+```tsx
+// pages/index.tsx
+
+export const getServerSideProps = async (context) => {
+  const res = await fetch("https://jsonplaceholder.typicode.com/posts");
+  const posts = await res.json();
+
+  return { props: { posts } };
+}
+
+export default function Page(props) {
+  return (
+    <div>
+      {props.posts.map(post => (
+        <div>
+          <h2>{post.title}</h2>
+          <p>{post.body}</p> 
+        </div>
+      ))}
+    </div>
+  )
+}
+```
+
+This is great for dynamic data that may not be best suited for `getStaticProps` such as fetching from a database or an API route with data that changes often. 
+
+The `context` parameter also has a lot of useful information about the request, including the request path, cookies sent from the client, and more that can be found on the [official Next.js documentation](https://nextjs.org/docs/pages/api-reference/functions/get-server-side-props#context-parameter).
+
+You can use [https://next-code-elimination.vercel.app/](https://next-code-elimination.vercel.app/) to verify what code is sent to the client when using `getServerSideProps`. 
+
+
+## Client Side Fetching
```

Comment:
I think we should add a note to be using `swr` or `react-query` or similar for api calls and moving away from calling api's inside of `useEffect` as this can lead to issues. (Slower rendering and multiple api calls) 

https://articles.wesionary.team/why-useeffect-is-a-bad-place-to-make-api-calls-98a606735c1c

Reply:
Yep, I plan to create a separate rule for this as both a React and Next.js rule as per my conversation with @jackreimers. I just included a useEffect fetch for simplicity - do you think it should be changed in this code snippet?

Reply:
@Harry-Ross I'd say we probably don't want to have bad practices in the code snippets, maybe change the snippet and once the new rule is written we can link to it to explain further?